server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobsDAO.java

```diff
@@ -122,71 +118,70 @@ private PreparedStatement prepareInsertPart(Session session) {
     public CompletableFuture<BlobId> save(byte[] data) {
         Preconditions.checkNotNull(data);
 
-        HashBlobId blobId = blobIdFactory.forPayload(data);
+        return saveAsMono(data).toFuture();
+    }
+
+    private Mono<BlobId> saveAsMono(byte[] data) {
+        BlobId blobId = blobIdFactory.forPayload(data);
         return saveBlobParts(data, blobId)
-            .thenCompose(numberOfChunk -> saveBlobPartsReferences(blobId, numberOfChunk))
-            .thenApply(any -> blobId);
+            .flatMap(numberOfChunk -> saveBlobPartsReferences(blobId, numberOfChunk));
     }
 
-    private CompletableFuture<Integer> saveBlobParts(byte[] data, HashBlobId blobId) {
-        return FluentFutureStream.of(
-            dataChunker.chunk(data, configuration.getBlobPartSize())
-                .map(pair -> writePart(pair.getRight(), blobId, pair.getKey())
-                    .thenApply(partId -> Pair.of(pair.getKey(), partId))))
-            .completableFuture()
-            .thenApply(stream ->
-                getLastOfStream(stream)
-                    .map(numOfChunkAndPartId -> numOfChunkAndPartId.getLeft() + 1)
-                    .orElse(0));
+    private Mono<Integer> saveBlobParts(byte[] data, BlobId blobId) {
+        Stream<Pair<Integer, ByteBuffer>> chunks = dataChunker.chunk(data, configuration.getBlobPartSize());
+        return Flux.fromStream(chunks)
+            .publishOn(Schedulers.elastic(), PREFETCH)
+            .flatMap(pair -> writePart(pair.getValue(), blobId, getChunkNum(pair)))
+            .collect(Collectors.maxBy(Comparator.comparingInt(x -> x)))
+            .flatMap(Mono::justOrEmpty)
+            .map(this::numToCount)
+            .defaultIfEmpty(0);
     }
 
-    private static <T> Optional<T> getLastOfStream(Stream<T> stream) {
-        return stream.reduce((first, second) -> second);
+    private int numToCount(int number) {
+        return number + 1;
     }
 
-    private CompletableFuture<Void> writePart(ByteBuffer data, HashBlobId blobId, int position) {
-        return cassandraAsyncExecutor.executeVoid(
+    private Integer getChunkNum(Pair<Integer, ByteBuffer> pair) {
+        return pair.getKey();
```

Comment:
I think it can be inlined

Reply:
to be honest, I think that having a method make the code easier to understand because `getKey` doesn't tell much